source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,80 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
+        //   go through IntCasts to deal with u32 -> i32 / vice-versa (IntCast)
+        // add the decoration to the first Use that is either an extract element  / a load of
+        // an extract element address instruction
+        // if the first use is an extract element ptr, find all the loads that uses the element,
+        // and the decorate them
+        //
+        // if the first use is a getelementptr, we may want to decorate the get element ptr itself,
+        // because user might
+        // call an image atomic operation, that operates on the ptr instead of the texture / resource
+        // todo rename the decoration itself nonuniformresourceindex -> nonuniformresource
+        List<IRInst*> resWorkList;
+        std::function<void(IRInst*)> visit;
+        visit = [&](IRInst* inst)
+        {
+            traverseUses(inst, [&](IRUse* use)
+            {
+                auto user = use->getUser();
+                IRBuilder builder(user);
+                builder.setInsertBefore(user);
+                IRInst* newUser = nullptr;
+                switch (user->getOp())
+                {
+                case kIROp_IntCast:
+                    printf("INTCAST!! %d\n", user->_debugUID);
+                    newUser = builder.emitCast(user->getFullType(), inst->getOperand(0));
+                    break;
+                case kIROp_GetElementPtr:
+                    // Replace gep(pArray, nonUniformRes(x)), into nonUniformRes(gep(pArray, x))
+                    printf("GETELEMENTPTR!! %d\n", user->_debugUID);
+                    newUser = builder.emitElementAddress(user->getFullType(), user->getOperand(0), inst->getOperand(0));
+                    break;
+                case kIROp_Load:
+                    printf("LOAD!! %d\n", user->_debugUID);
+                    newUser = builder.emitLoad(user->getFullType(), inst->getOperand(0));
+                    break;
+                default:
+                    printf("UNKNOWN!! %d\n", user->_debugUID);
+                    break;
+                };
+                if (newUser)
+                {
+                    auto nonUniformUser = builder.emitNonUniformResourceIndex(newUser->getFullType(), newUser);
+                    user->replaceUsesWith(nonUniformUser);
+                    resWorkList.add(nonUniformUser);
+                    //visit(nonUniformUser);
+                }
+            });
+        };
+        visit(nonUniformResourceIndexInst);
+
+        // Now we have "upstream"ed all `NonUniformResourceIndex` insts, 
+        // the next step is to add a decoration the operands of all nonuniformresourceindex insts
+        for (auto nonUniformResInst : resWorkList)
+        {
+            auto operand = nonUniformResInst->getOperand(0);
+            IRBuilder builder(nonUniformResInst);
+            //if (isResourceType(operand->getDataType())) // || operand's type is a pointer to resource type)
```

Comment:
I think we still need to guard the type so we don't add the decoration on unrelated things. isResourceType probably didn't include the buffer types, need to double check.

Reply:
isResourceType is not available in spirv-legalize. could you please help me understand the meaning of this requirement?

Reply:
Find the function and maybe declare it in slang-ir-util.h so you can use it here. The purpose is to avoid decorating on things that are irrelevant. For example no need to declare an int value as non uniform.